src/externs.ts

```diff
@@ -78,24 +79,88 @@ export function getGeneratedExterns(externs: {[fileName: string]: string}): stri
  * file. It returns a string representation of the Closure JavaScript, not including the initial
  * comment with \@fileoverview and \@externs (see above for that).
  */
 export function generateExterns(
-    typeChecker: ts.TypeChecker, sourceFile: ts.SourceFile,
-    host: AnnotatorHost): {output: string, diagnostics: ts.Diagnostic[]} {
+    typeChecker: ts.TypeChecker, sourceFile: ts.SourceFile, host: AnnotatorHost,
+    moduleResolutionHost: ts.ModuleResolutionHost,
+    options: ts.CompilerOptions): {output: string, diagnostics: ts.Diagnostic[]} {
   let output = '';
   const diagnostics: ts.Diagnostic[] = [];
   const isDts = isDtsFileName(sourceFile.fileName);
   const mtt =
       new ModuleTypeTranslator(sourceFile, typeChecker, host, diagnostics, /*isForExterns*/ true);
 
+  let namespace: string[] = [];
+  let rootNamespace = '';
+  const isExternalModule = ts.isExternalModule(sourceFile);
+  if (isExternalModule) {
+    // .d.ts files that are modules do not declare global symbols - their symbols must be
+    // explicitly imported to be used. However Closure Compiler has no concept of externs that
+    // are modules and require imports. This code hides the global symbols from code by
+    // wrapping them in a top level variable that's unique to this file.
+    // This is necessary as symbols local to this can (and will very commonly) conflict with
+    // the namespace used in "export as namespace", e.g. "angular", and also to avoid users
+    // accidentally using these symbols in .js files (and more collisions).
+    // The symbols that are "hidden" like that can be made accessible through an "export as
+    // namespace" declaration (see below).
+    rootNamespace = mangledModuleName(host, sourceFile.fileName);
+    namespace = [rootNamespace];
+  }
+
   for (const stmt of sourceFile.statements) {
     if (!isDts && !hasModifierFlag(stmt, ts.ModifierFlags.Ambient)) continue;
-    visitor(stmt, []);
+    visitor(stmt, namespace);
+  }
+
+  if (output && isExternalModule) {
+    // If tsickle generated any externs and this is an external module, prepend the namespace
+    // declaration for it.
+    output = `/** @const */\nvar ${rootNamespace} = {};\n` + output;
+
+    // There can only be one export =.
+    const exportAssignment = sourceFile.statements.find(ts.isExportAssignment);
+    let exportedNamespace = rootNamespace;
+    if (exportAssignment && exportAssignment.isExportEquals) {
+      if (ts.isIdentifier(exportAssignment.expression) ||
+          ts.isQualifiedName(exportAssignment.expression)) {
+        exportedNamespace = rootNamespace + '.' + getEntityNameText(exportAssignment.expression);
+      } else {
+        error(exportAssignment.expression, `export = expression must be a qualified name.`);
+      }
+    }
+
+    if (isDts && host.provideExternalModuleDtsNamespace) {
+      const closureNamespace = host.pathToModuleName('', sourceFile.fileName);
```

Comment:
Will this include the 'google3.' prefix (when run outside of the test suite)?

Reply:
Yes, this is the function we use to generate all module names.